src/legacy/core_plugins/telemetry/server/get_telemetry_opt_in.ts

```diff
@@ -43,4 +57,50 @@ export async function getTelemetryOptIn(request: any) {
 
     throw error;
   }
+
+  const { attributes }: { attributes: SavedObjectAttributes } = savedObject;
+
+  // if enabled is already null, return null
+  if (attributes.enabled == null) return null;
+
+  const enabled = !!attributes.enabled;
+
+  // if enabled is true, return it
+  if (enabled === true) return enabled;
+
+  // Additional check if they've already opted out (enabled: false):
+  // - if the Kibana version has changed by at least a minor version,
+  //   return null to re-prompt.
+
+  const lastKibanaVersion = attributes.lastVersionChecked;
+
+  // if the last kibana version isn't set, or is somehow not a string, return null
+  if (typeof lastKibanaVersion !== 'string') return null;
+
+  // if version hasn't changed, just return enabled value
+  if (lastKibanaVersion === currentKibanaVersion) return enabled;
+
+  const lastSemver = parseSemver(lastKibanaVersion);
+  const currentSemver = parseSemver(currentKibanaVersion);
+
+  // if either version is invalid, return null
+  if (lastSemver == null || currentSemver == null) return null;
```

Comment:
for cases like this, where there's a local function I can validate won't return `undefined`, I'm happier to to use a `===` check, but still kind of consider it to be bad form, for future-proofing in case some day the called function DOES end up returning `undefined`.

I'm curious why you're looking for this.

Reply:
The docs for semver's `parse` specifically state that null is returned if the value can't be parsed. You're guessing that in the future, if `parse` starts to return `undefined`, that `undefined` will mean the same as `null`. But that is not necessarily the case. Some future version of `parse` might return `undefined` for some other reason, and then we might have a bug on our hands because this condition is overly loose. I don't consider `null` and `undefined` interchangeable, they can mean two very different things depending on context.

Reply:
> Some future version of `parse()` might return `undefined` for some other reason, and then we might have a bug on our hands because this condition is overly loose.

I guess I'd say, based on my years-long, unhappy history with `semver` (as much as I don't like it, it gets a lot of use, so it's the best out there, as long as you protect yourself), that we'd more likely see some future of version of `parse()` return `undefined` where it returns `null` today, and then we might have a bug on our hands because `=== null` is overly strict.  :-)

I don't trust 3rd party libraries much ...

Reply:
I think it's also the case that I've used linters (`standard`?) that used to complain if you ever DID do an `=== null` or `=== undefined`.  Happened to fit with my world view, so was happy with it.  Or was I brainwashed?!?! heh

Reply:
> that we'd more likely see some future of version of parse() return undefined where it returns null today, and then we might have a bug on our hands because === null is overly strict. :-)

Nope, because our unit tests would catch this. The `=== null` condition would fail and we would get an unexpected result. This is the benefit of being strict. On the other hand, our unit tests would likely not catch the bug in my example because the looseness of the check may cover up an issue. Semver might respond to some issue by returning undefined but this `==` check will make it look like everything is fine. 

> my years-long, unhappy history with semver

My decades long, unhappy history with Javascript has taught me to never use double equals :)   It's not just me:

https://2ality.com/2011/12/strict-equality-exemptions.html (look especially at Case 2)

https://stackoverflow.com/a/359509 (note in particular the quote from Javascript: The Good Parts)

Preferring triples equals is a best practice most of the community agrees upon. 

But like I said above, the stakes in this particular situation are probably low, so if you are not convinced, we can agree to disagree.

Reply:
pro-tip: you will never convince me of anything based on references to "JavaScript: The Good Parts". hehe

I'm slightly curious, because I thought triple quotes were best practice for number/boolean/string, but double quotes was actually best practice for null/undefined.

It would certainly be interesting to add a lint rule disallowing `== null` and see what the vsCode Problems page looks like.

quick search:

search | results | files
-- | -- | --
` == null` | 600 | 236
` === null` | 339 | 177
` === undefined` | 573| 284
`== undefined` | 8 | 1 hahahaha

Reply:
I was attempting an appeal to authority out of frustration since the concrete explanations I gave you are for some reason unconvincing. You didn't actually respond to the concrete explanation I gave about this specific piece of code we're looking at. You also ignored the information from the first link I provided, which has a section (which I called out) that specifically deals with this particular use of `==`. Instead you made a flippant comment about Javascript: The Good Parts and listed a statistic from the existing code base which is irrelevant to the discussion of whether using `==` is actually a good idea, or if it is appropriate to use in this specific context. You also keep noting that the default ruleset of some linter you used encouraged use of `==`. This is also an appeal to authority, except the authority is anonymous and their reasoning is unknown, so I would say this is also irrelevant to this specific conversation. 

I won't reply anymore, because I'm obviously extremely frustrated and this is not a productive conversation.

Reply:
Note: sent a note to Bargs offline, as I didn't realize his frustration level on this - I thought the issue was resolved but we were just having some banter on the topic of `== null`.